yourbench/pipeline/answer_generation.py

```diff
@@ -0,0 +1,276 @@
+            scenario=scenario,
+            output_subset="multi_hop_questions_with_answers",
+        )
+
+    logger.success("answer_generation stage complete.")
+
+
+def _generate_answers_for_questions(
+    config: Dict[str, Any], source_subset: str, scenario: str, output_subset: str
+) -> None:
+    """
+    Loads a question subset (single_shot or multi_hop), runs the chosen scenario
+    (zero_shot or with_correct_chunk), and appends results to the specified
+    output_subset, storing 'scenario' in answer_fashion.
+
+    The final dataset has new columns:
+      - 'answering_model'
+      - 'answer'
+      - 'answer_fashion' = scenario
+    """
+    logger.info(f"Loading questions from '{source_subset}' for scenario='{scenario}'.")
+    try:
+        question_ds = custom_load_dataset(config, subset=source_subset)
+    except Exception as e:
+        logger.warning(f"Could not load {source_subset} dataset. Error: {e}")
+        return
+
+    if question_ds is None or len(question_ds) == 0:
+        logger.warning(f"No data in {source_subset}; skipping answer generation.")
+        return
+
+    # Build inference calls for the chosen scenario
+    calls, row_map = _build_inference_calls_scenario(config, question_ds, scenario, source_subset)
+    if not calls:
+        logger.warning(f"No inference calls for subset={source_subset}, scenario={scenario}.")
+        return
+
+    # Run them
+    responses = run_inference(
+        config=config,
+        step_name="answer_generation",  # "model_roles" can link here
+        inference_calls=calls,
+    )
+    if not responses:
+        logger.warning(f"No responses returned for subset={source_subset}, scenario={scenario}.")
+        return
+
+    # Parse and assemble
+    final_ds = _parse_and_assemble(
+        question_ds=question_ds, responses_dict=responses, row_map=row_map, scenario=scenario
+    )
+    if final_ds is None or len(final_ds) == 0:
+        logger.warning(f"No answers parsed for subset={source_subset}, scenario={scenario}.")
+        return
+
+    # Append to the same subset
+    custom_save_dataset(dataset=final_ds, config=config, subset=output_subset)
+    logger.info(f"Appended {len(final_ds)} new answers to '{output_subset}' (scenario={scenario}).")
+
+
+def _build_inference_calls_scenario(config: Dict[str, Any], question_ds: Dataset, scenario: str, source_subset: str):
+    """
+    For each row in question_ds, build an InferenceCall with the appropriate prompt:
+     - "zero_shot" => use ZEROSHOT_QA_USER_PROMPT
+     - "with_correct_chunk" => retrieve chunk(s) from 'chunked' dataset and use GOLD_QA_USER_PROMPT
+
+    Returns (calls, row_map) for re-assembly after inference.
+    """
+    calls = []
+    row_map = []
+
+    if "question" not in question_ds.column_names:
+        logger.error("Dataset missing 'question' column. Cannot build calls.")
+        return calls, row_map
+
+    doc_meta_map = {}
+    if scenario == "with_correct_chunk":
```

Comment:
You need to split here the if, and iterate on the rows inside the clauses - else you're adding an extra if check at each new row, it's an extra operation which is useless.

```python
if scenario == "zero shot":
    for row;
        do the thing
elif scenario == "with correct chunks":
    load dataset
    for row:
        do the thing
else:
 etc
````

Reply:
I'll re do a review once this is fixed

Reply:
The `_build_inference_calls_scenario()` function now has a top-level branch on `scenario` `(if scenario == "zero_shot": ... elif scenario == "with_correct_chunk": ...)`, so we avoid repeated scenario checks within the row loop.